### ISSUE TITLE
missing testing for cl_ext_float_atomics

### ISSUE THREAD

**Issue by bashbaug** [KhronosGroup/OpenCL-CTS#2142]
This issue is to track missing tests for the `cl_ext_float_atomics` extension.

There is some testing for these extensions in the Vulkan CTS.

There are also some tests in the Intel Cassian test framework, see: https://github.com/intel/cassian/commit/239de5d8630bee09c0f3e136d4dab6e508bf80e3

A good first step would be to evaluate what is tested by the other frameworks, see what may be useful for the OpenCL CTS, and write up a brief test plan.

**Comment by shajder** [comment]
Below is the proposition for a work plan on `cl_ext_float_atomics` and a few questions, I will appreciate discussions and your answers, thanks!

1. Adaptation of `c11_atomics` test suite in order to add support for `cl_ext_float_atomics` by including new generic cases for data types `fp16`/`fp32`/`fp64` in classes:

- [x] [CBasicTestStore](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L27) with atomic_half
- [x] [CBasicTestLoad](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L295) with atomic_half
- [x] [CBasicTestExchange](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L470) with atomic_half
- [x] [CBasicTestFetchAdd](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L1055) with atomic_float
- [x] [CBasicTestFetchAdd](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L1055) with atomic_double
- [x] [CBasicTestFetchAdd](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L1055) with atomic_half
- [x] [CBasicTestFetchSub](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L1197) with atomic_float
- [x] [CBasicTestFetchSub](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L1197) with atomic_double
- [x] [CBasicTestFetchSub](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L1197) with atomic_half
- [x] [CBasicTestFetchMin](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L2121) with atomic_float
- [x] [CBasicTestFetchMin](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L2121) with atomic_double
- [x] [CBasicTestFetchMin](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L2121) with atomic_half
- [x] [CBasicTestFetchMax](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L2256) with atomic_float
- [x] [CBasicTestFetchMax](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L2256) with atomic_double
- [x] [CBasicTestFetchMax](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L2256) with atomic_half

2. Extending c11_atomics with special values tests for:

- [x] [CBasicTestFetchAdd](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L1055) with atomic_float
- [x] [CBasicTestFetchAdd](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L1055) with atomic_double
- [x] [CBasicTestFetchAdd](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L1055) with atomic_half
- [x] [CBasicTestFetchSub](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L1197) with atomic_float
- [x] [CBasicTestFetchSub](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L1197) with atomic_double
- [x] [CBasicTestFetchSub](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L1197) with atomic_half
- [x] [CBasicTestFetchMin](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L2121) with atomic_float
- [x] [CBasicTestFetchMin](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L2121) with atomic_double
- [x] [CBasicTestFetchMin](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L2121) with atomic_half
- [x] [CBasicTestFetchMax](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L2256) with atomic_float
- [x] [CBasicTestFetchMax](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L2256) with atomic_double
- [x] [CBasicTestFetchMax](https://github.com/KhronosGroup/OpenCL-CTS/blob/9f63decb9c85875c3f285005a508037233f530b9/test_conformance/c11_atomics/test_atomics.cpp#L2256) with atomic_half

3. Implementation of new tests for `spirv_new` test suite including both `CrossWorkGroup` and `Workgroup` storage classes, supporting `fp16`/`fp32`/`fp64` is not necessary. It is a matter of `--compilation-mode spir-v` run options and the specific `cl_offline_compiler` to test SPIRV + `cl_ext_float_atomics` extension along with `c11_atomics` test.

**Comment by bashbaug** [comment]
IMHO, one of the trickiest parts of testing float atomics will be numerical stability because floating-point addition is not associative.  Have you given any thought how to handle this issue?  Have you evaluated how other testing frameworks solve this problem?

Are we able to test any NaN (and infinity?) behavior, or are we expecting our inputs and results to be non-NaN (and non-infinity)?

**Comment by b-sumner** [comment]
I agree with @bashbaug.  Unless the order of the atomic operations is explicitly controlled, the test will have to assume they can happen in any possible order and account for that.

An option for one test is to use only the values where the order doesn't matter and each operation is guaranteed to be exact, e.g. small integers while avoiding overflow.

**Comment by karolherbst** [comment]
I noticed that the compiler test isn't testing for the feature macros, should probably add this

**Comment by shajder** [comment]
> I noticed that the compiler test isn't testing for the feature macros, should probably add this

@karolherbst regarding your comment about the compiler test not checking the feature macros: This issue will be closed soon, since all planned work is done, so the note would get lost. Would you like to open a separate issue for it, or should I do it and reference your comment?

**Comment by karolherbst** [comment]
> > I noticed that the compiler test isn't testing for the feature macros, should probably add this
> 
> [@karolherbst](https://github.com/karolherbst) regarding your comment about the compiler test not checking the feature macros: This issue will be closed soon, since all planned work is done, so the note would get lost. Would you like to open a separate issue for it, or should I do it and reference your comment?

You can create a new issue for it, thanks.
